src/components/dialog/use-dialog-appearance.ts

```diff
@@ -0,0 +1,31 @@
+import { RefObject, useEffect, useRef } from 'react'
+
+interface UseDialogAppearanceOptions {
+  isOpen: boolean
+  onClose: () => void
+}
+
+export const handleDialogAppearance =
+  (options: UseDialogAppearanceOptions, ref: RefObject<HTMLDialogElement>) => () => {
+    const { isOpen, onClose } = options ?? {}
+    if (!ref.current || !onClose) return
+
+    if (isOpen) {
+      ref.current.showModal()
+      ref.current.addEventListener('close', onClose)
+    } else {
+      ref.current.removeEventListener('close', onClose)
+      ref.current.close()
+    }
+
+    return () => {
+      ref.current?.removeEventListener('close', onClose)
+    }
+  }
+export const useDialogAppearance = ({ isOpen, onClose }: UseDialogAppearanceOptions) => {
```

Comment:
having a solid foundation for this hook is great, agree all of your thoughts here

for your third thought, the `useToggleDialogVisibilityEffect` is more understandable to me

**question:** shall we also need to have a callback for both scenario? like the way on `onClose` did in `useDialogAppearance` at this moment

Reply:
> question: shall we also need to have a callback for both scenario? like the way on onClose did in useDialogAppearance at this moment

Yes, sorry, it would still accept the `onClose` callback like it currently does.